_posts/2019/2019-01-13-use-multiple-migrations-when-adding-database-constraints.md

```diff
@@ -0,0 +1,95 @@
+---
+title: Use Multiple Migrations When Adding Database Constraints
+description: You need to avoid a long table rebuild and a write-lock
+layout: article
+category: ruby
+image:
+  base: '2018/use-multiple-migrations-when-adding-database-constraints'
+  alt: 'Locks'
+  source: 'https://unsplash.com/photos/8_NI1WTqCGY'
+  credit: 'marcos mayer'
+---
+
+Adding constraints to your application at the database layer is a good idea as it gives an extra layer of quality control to the data powering your application.
+
+Adding or changing default values, or even adding a new column with a default value to an existing table, is easy to do with Rails’ migrations. But beware...
+
+
+## Instead of…
+
+…adding a column with a default non-null value in one go.
+
+```ruby
+class AddComposerToSongs < ActiveRecord::Migration[5.2]
+  def change
+    add_column :songs, :composer, :string, default: "Lin-Manuel Miranda", null: false
+  end
+end
+```
+
+
+## Use…
+
+…a multiple migration strategy to add a field with a constraint to your databases
+
+
+### Add a column
+
+```ruby
+class AddComposerToSongs < ActiveRecord::Migration[5.2]
+  def change
+    add_column :songs, :composer, :string
+  end
+end
+```
+
+
+### Deploy code that sets default, in Ruby
+
+```ruby
+class Song < ApplicationRecord
+  before_save :set_default_composer
+
+  def set_default_composer
+    self.composer = "Lin-Manuel Miranda" if composer.nil?
+  end
+end
+```
+
+
+### Update records with nil values
+
+```ruby
+Song.where(composer: nil).update_all(composer: "Lin-Manuel Miranda")
+```
+
+
+### Change the column to have a constraint
+
+```ruby
+class AddRecommendedIndexes < ActiveRecord::Migration[5.1]
+  def change
+    change_column :songs, :composer, :string, default: "Lin-Manuel Miranda", null: false
+  end
+end
+```
+
+…and then you can delete the Ruby code that sets a default.
+
+
+## But why?
+
+If you take the “naive” approach you run the risk of significant downtime for your application.
```

Comment:
Not sure if labelling it 'naive' is fair, or the tone you want to go for. After all, people don't know — and you're enlightening them! Maybe 'all-in-one' approach?